item19. 상속을 고려해 설계하고 문서화하라. 그러지 않았다면 상속을 금지하라

Limwin94 opened the pull request:
- 상속을 고려한 설계와 문서화
    - 부작용
- 문서화(자세히)
    - Java
    - Swift
- 상속에 대한 설계
    - 어떤 메소드를 재정의 할 수 있게 해야하나?
    - 제약사항
- 상속용 클래스가 아닌 일반적인 클래스에 대한 경우

이렇게 정리해보았습니다.

Limwin94 committed:
item19 작성

Limwin94 committed:
update item19.md

상속에 대한 설계
 - 제약사항에 대한 내용 수정

delmaSong commented on 4장_클래스와_인터페이스/item19.md:
```diff
@@ -0,0 +1,107 @@
+
+<br>
+
+### 상속을 고려한 설계와 문서화
+- 문서화를 할때 고려해야 할 부분들 입니다.
+    1. 상속용 클래스(상위 클래스)는 재정의 할 수 있게 만든 메서드들을 내부적으로 어떻게 사용하는지 문서로 남겨야 합니다.
+    2. 공개(public, open) 메서드에서 재정의 가능한 메서드를 호출 할 때, API에 명시하여야 합니다.
+    3. 어떤 순서로 호출하는지, 각각의 호출 결과가 이어지는 처리에 어떤 영향을 주는지도 담아야 합니다.
+    
+#### 부작용
+클래스를 안전하게 상속할 수 있도록 문서화를 해놓는 것을 권장하지만, 캡슐화를 해치게 된다는 부작용이 있습니다. '무엇'을 하는지만 명시하는게 아니라 이 클래스 혹은 메서드가 '어떻게' 동작하는 지도 설명해야하기 때문입니다.
+
+<br>
+
+### 문서화
+#### Java
+자바의 경우 API 문서에서는 Implementation Requirements로 시작하는 절이 있는데, 해당 절은 @implSpec 태그를 붙여주면 자바독 도구가 생성해줍니다.
+> Javadoc : Java 소스를 문서화 하는 방법, html로 열 수 있습니다.
```
`상속을 고려한 설계와 문서화`를 꼭지로 `문서화`와 합칠 수 있을 것 같아요. ㅎㅎ

Limwin94 replied:
수정하였습니다~!

delmaSong commented on 4장_클래스와_인터페이스/item19.md:
````diff
@@ -0,0 +1,107 @@
+예제에서는 Java의 `removeRange` 메서드를 예로 들고있습니다.
+해당 메서드를 protected로 제공한 이유는 `clear` 메서드가 이 함수를 부르기 때문인데.
+        
+```Java
+protected void removeRange(int fromIndex, int toIndex) {
+    ListIterator<E> it = listIterator(fromIndex);
+    for (int i=0, n=toIndex-fromIndex; i<n; i++) {
+        it.next();
+        it.remove();
+    }
+}
+```
+`removeRange`메서드 내부의 `it.remove()`가 O(n)이 걸리기 때문에 실질적으로 `removeRange`는 O(n^2)이 걸리게 되어버립니다.
+따라서 해당 메서드를 호출하는 `clear` 메서드를 고성능으로 만들기 쉽게 하기 위해 `removeRange` 메서드를 재정의 할수있도록 제공하는 것입니다. 
````
책을 읽지 않으면 해당 예시를 왜 들었는지 이해하기 어려운 것 같아요ㅠ  설명을 조금 보충해주시면 좋겠습니다 
그리고 스위프트에 대한 예시도 있으면 .. 좋을 것 같습니다 ㅎㅎ

seizze replied:
이건 진짜 사소한 거일수도 있는데.. 책에서는 **`it.remove()`가 O(n)이 걸린다면** `removeRange`는 O(n^2)이 걸린다고 나와 있어서요. 항상 remove()가 O(n)이 걸린다는 건 아닌 것 같아요~

Limwin94 replied:
@delmaSong 현재는 설명을 조금 더 보충하는 방안으로 가볼게요, swift에 대한 예시는 오래 찾아봐도 찾기 힘들어서 추후 발견하면 추가하는 방향으로 잡아볼게요. 의견감사합니다.

@seizze 그러네요, 문구 수정하겠습니다 감사합니다.

delmaSong reviewed (approved):
수고하셨습니다 👏🏼

seizze commented on 4장_클래스와_인터페이스/item19.md:
```diff
@@ -0,0 +1,107 @@
+
+<br>
+
+### 상속에 대한 설계
```
"상속을 위한 설계"가 더 적합할 것 같은데 어떻게 생각하시나요?

Limwin94 replied:
상속을 위한 설계 / 문서화로 변경하였습니다. 의견 감사합니다.

seizze commented on 4장_클래스와_인터페이스/item19.md:
```diff
@@ -0,0 +1,107 @@
+
+- 요지는 Java에서는 상위 클래스의 생성자가 먼저 호출되어 하위 클래스의 overrideMe 메소드를 부르게 되는데 이때 subProperty가 초기화되어있지 않은 상태에서 부르게 되므로 `NullPointerException`을 던지게 된다는 것이다.
+- Swift에서는 상위 클래스 생성자를 호출하기 위해서는 하위 클래스의 stored property에 모든 값이 초기화 된 후 호출할 수 있기 때문에 예시와 같은 오류가 발생하기 어렵지만, stored property가 optional일 경우에는 위와 같은 상황이 발생할 여지가 있다. 
+- 다만 개인적인 생각으로는 재정의 가능한 함수를 생성자에서 호출시킨다는 것은 예측가능하기 어려운 흐름이 발생할 수 도 있기 때문에 지양하는 방향이 올바르다고 생각합니다.
```
ㅎㅎ 합니다체로 통일부탁드립니다 ☺️

seizze replied:
> Swift에서는 상위 클래스 생성자를 호출하기 위해서는 하위 클래스의 stored property에 모든 값이 초기화 된 후 호출할 수 있기 때문에 예시와 같은 오류가 발생하기 어렵지만, stored property가 optional일 경우에는 위와 같은 상황이 발생할 여지가 있다. 

이 말은 이해가 잘 안가는데, stored property가 옵셔널일 경우엔, 상위 클래스의 생성자가 먼저 호출되어 하위 클래스의 overrideMe 메소드를 부르게 되어 나타나는 문제점과 연관이 없는거 같아요.. 어떻게 생각하시나요?

seizze replied:
그리고 저는 스위프트에서 **상위 클래스 생성자를 호출하기 위해서는 하위 클래스의 stored property에 모든 값이 초기화 된 후 호출**하도록 강제한 건 코드의 안정성을 높이기 위해서 내린 설계 방식이라는 생각이 들었는데요, 그래서 스위프트 프로그래밍 랭귀지 사이트를 찾아봤는데 Class Inheritance and Initialization 파트에 Two-Phase Initialization이라는 내용이 있더라고요~

저도 훑어 보기만 했는데, 뭔가 상속 관계의 두 클래스를 초기화할 때 기존 언어에선 여러 문제점들이 있는데(아마 이번 아이템도 그 문제점들 중 하나라고 추측됩니다), 이를 해결하기 위해 뭔가 초기화 단계를 두 단계로 나누어 안전성을 여러 번 체크하는 느낌이었습니다.

내용이 많아서 추가하라는 말씀은 못드리겠고, 읽고 싶으면 한번 읽어보세요 ㅎㅎ
https://docs.swift.org/swift-book/LanguageGuide/Initialization.html#ID216

Limwin94 replied:
@seizze 간단하게 읽어보았는데요, 도움되는 내용인것 같아 추가하려고 합니다만...
통째로 떼어서 다루어야 할 내용인것 같아 간단하게 언급만 하고 넘어가려고 합니다.

문서 자체는 잘 읽어보았습니다.

감사합니다~!

seizze reviewed (changes requested):
수고하셨습니다~ 리뷰가 쫌 많은데, 랭귀지 가이드 관련 코멘트는 수정 요청은 아닙니다 ㅎㅎ

Limwin94 committed:
문서화 내용 수정

상속을 고려한 설계와 문서화에 대한 목차에서 세부 내용이 '상속을 고려한 설계'에 대한 내용을 포함하고 있지 않고 밑에서 설명하기 때문에 문서화의 목차만 남겨놓았습니다.

Limwin94 committed:
예시중 removeRange 메서드에 대한 내용 수정

피드백 내용 반영 및 문장이 매끄러울 수 있도록 수정했습니다.

Limwin94 committed:
문구 수정

상속'에 대한' -> 상속'을 위한'으로 수정하였습니다.

Limwin94 committed:
문체 통일

-다. -> -합니다. 체로 통일하였습니다.

Limwin94 committed:
내용 보충

Swift 초기화 흐름에 대한 짤막한 설명 추가, 글을 작성하면서 참고한 사이트 링크 추가

dev-Lena reviewed (approved):
자바와 스위프트에서의 상속에 대해 잘 봤습니다!
수고하셨습니다 👏🏻

seizze reviewed (approved):
수고하셨습니다~~👏👏👏

ehgud0670 commented on 4장_클래스와_인터페이스/item19.md:
```diff
@@ -0,0 +1,113 @@
+    3. 어떤 순서로 호출하는지, 각각의 호출 결과가 이어지는 처리에 어떤 영향을 주는지도 담아야 합니다.
+    
+#### 부작용
+클래스를 안전하게 상속할 수 있도록 문서화를 해놓는 것을 권장하지만, 캡슐화를 해치게 된다는 부작용이 있습니다. '무엇'을 하는지만 명시하는게 아니라 이 클래스 혹은 메서드가 '어떻게' 동작하는 지도 설명해야하기 때문입니다.
```
"좋은 API 문서는 어떻게가 아닌 무엇을 하는지를 설명해야 한다" 라는 격언과 대치되는 것이 부작용이지, 문서화를 통해 캡슐화를 해친다는 말은 아닌 것 같습니당.
상속이 애초에 캡슐화를 해치는 것이고, 상속 때문에 좋은 API문서의 정의대로 설명하지 못하는게 부작용인 것 같습니다!

(https://thdev.tech/kotlin/2020/12/01/kotlin_effective_13/  => 이 링크를 보고 코멘트 남깁니다.)

ehgud0670 commented on 4장_클래스와_인터페이스/item19.md:
````diff
@@ -0,0 +1,113 @@
+#### 제약사항
+상속용 클래스의 생성자는 직접적으로든 간접적으로든 재정의 가능 메서드를 호출해서는 안됩니다.
+
+```Swift
+class Super {
+    init() {
+        overrideMe()
+    }
+            
+    public func overrideMe() { }
+}
+        
+class Sub {
+    private var subProperty: Int
+            
+    init() {
+        self.subProperty = 100    
+    }
+            
+    override func overrideMe() {
+        print(subProperty)
+    }
+}
+        
+let sub = Sub()
+sub.overrideMe()
+```
````
=> 자바 코드를 스위프트 코드로 바꾸신것 같은데 상속관계가 드러나지 않아서 아래와 같이 제안합니다.
그리고 아래 코드는 아래 말씀하신것처럼 stored property가 optional 한 프로퍼티인 경우이기 때문에 자바처럼 런타임에러가 발생하므로 더욱이 설명하기에 적절한 코드라고 생각해 제안드립니다~!!

```suggestion
```Swift
class Super {
    init() {
        overrideMe()
    }
            
    public func overrideMe() { }
}
        
class Sub: Super {
    private var subProperty: String!
            
    init(subProperty: String?) {
        self.subProperty = subProperty
    }
            
    override func overrideMe() {
        print(subProperty.description)
    }
}

let sub = Sub(subProperty: nil) // 런타임 에러 발생! Thread 1: Fatal error: Unexpectedly found nil while implicitly unwrapping an Optional value
sub.overrideMe()
```
```

Limwin94 replied:
엇 왜 `Sub` 클래스가 `Super` 클래스를 상속안했는지 모르겠네요ㅋㅋ 옮기는 와중에 놓쳤나봅니다ㅠ

제안 감사합니다 :)

ehgud0670 reviewed (changes requested):
좋은 내용 잘 읽었습니다. 수고하셨습니다! 👏

Limwin94 committed:
Update 4장_클래스와_인터페이스/item19.md

상속 관계가 명확히 나타내도록 변경.

Co-authored-by: Doehyung Kim <kimdo2297@gmail.com>

Limwin94 committed:
단어 통일(메소드 -> 메서드)

Limwin94 committed:
문서화의 부작용에 대한 설명을 수정.

문서화로 인한 캡슐화의 실패가 아닌 상속 자체에 의한 캡슐화의 실패이다. (Jason 피드백)

Limwin94 committed:
코드 표시선 추가

ehgud0670 commented on 4장_클래스와_인터페이스/item19.md:
```diff
@@ -0,0 +1,114 @@
+    3. 어떤 순서로 호출하는지, 각각의 호출 결과가 이어지는 처리에 어떤 영향을 주는지도 담아야 합니다.
+    
+#### 부작용
+좋은 API 문서는 API가 '어떻게' 동작하는지 아닌 '무엇'을 하는지만을 설명해야 하지만, 상속으로 인해 캡슐화를 해치게 되어 내부 구현 방식을 설명해야하기 때문입니다.
```
```suggestion
문서화할 때 '어떻게' 동작하는지도 설명해야 하는 것이 부작용입니다. 좋은 API 문서는 API가 '어떻게' 동작하는지 아닌 '무엇'을 하는지만을 설명해야 하지만, 상속으로 인해 캡슐화를 해치게 되어 내부 구현 방식을 설명해야하기 때문입니다.
```
=> 위처럼 앞에 문장 추가 제안드립니다 ㅎㅎ

ehgud0670 reviewed (commented):
고생하셨습니다!! 새로 추가한것만 살펴보시고 코멘트 다시면 approve 하겠습니다. 잘 읽었습니다 👍 :)

Limwin94 committed:
부작용 목차 두괄식 표현으로 변경

Co-authored-by: Doehyung Kim <kimdo2297@gmail.com>

Limwin94 commented:
> 고생하셨습니다!! 새로 추가한것만 살펴보시고 코멘트 다시면 approve 하겠습니다. 잘 읽었습니다 👍 :)

변경해보았습니다. 제안 감사드려요 :)

ehgud0670 reviewed (approved):
수고하셨습니다!! 🎉
